### ISSUE TITLE
Using running mean in forward pass when model is parallelized

### ISSUE THREAD

**Issue by johnwlambert** [pytorch/pytorch#1500]
Hi, in my _init_ function of my nn.Module I define variables for a running mean. I then parallelize the model with data_parallel and place the model onto the GPUs I have with .cuda(). I use the running mean during my forward pass, but encounter an error telling me that tensors are located on different GPUs. What's the best way to overcome this sort of obstacle in the PyTorch paradigm? Making only certain parts of my forward pass parallelized, and calling .cuda(0) before the section with the running mean to place everything on GPU 0, for example?

Are there any good examples for this? I assume that you all did something similar when computing running means and variances for your batch norm implementation but couldn't find anything easily.

**Comment by apaszke** [comment]
If you want your buffers to be sent to the GPU along with all the module parameters, you should save them as Tensors and use the `register_buffer` method. See e.g.  [BatchNorm source](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L23-L24).

Note that we're using GitHub issues for bug reports only, and all questions should be posted on [our forums](discuss.pytorch.org).

**Comment by ngimel** [comment]
Note also that batch norm does not correctly handle averaging running variables with DataParallel, because updates from non-master GPUs are discarded #1051.
